core/src/main/java/org/elasticsearch/cluster/routing/DelayedAllocationService.java

```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Override
+    protected void doStop() {
+    }
+
+    @Override
+    protected void doClose() {
+        clusterService.remove(this);
+        removeTaskAndCancel();
+    }
+
+    /** override this to control time based decisions during delayed allocation */
+    protected long currentNanoTime() {
+        return System.nanoTime();
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        long currentNanoTime = currentNanoTime();
+        if (event.state().nodes().isLocalNodeElectedMaster()) {
+            scheduleIfNeeded(currentNanoTime, event.state());
+        }
+    }
+
+    private void removeTaskAndCancel() {
+        DelayedRerouteTask existingTask = delayedRerouteTask.getAndSet(null);
+        if (existingTask != null) {
+            logger.trace("cancelling existing delayed reroute task");
+            existingTask.cancelScheduling();
+        }
+    }
+
+    private void removeIfSameTask(DelayedRerouteTask expectedTask) {
+        delayedRerouteTask.compareAndSet(expectedTask, null);
+    }
+
+    /**
+     * Figure out if an existing scheduled reroute is good enough or whether we need to cancel and reschedule.
+     */
+    private void scheduleIfNeeded(long currentNanoTime, ClusterState state) {
+        assertClusterStateThread();
+        long nextDelayNanos = UnassignedInfo.findNextDelayedAllocation(currentNanoTime, state);
+        if (nextDelayNanos < 0) {
+            logger.trace("no need to schedule reroute - no delayed unassigned shards");
+            removeTaskAndCancel();
+        } else {
+            TimeValue nextDelay = TimeValue.timeValueNanos(nextDelayNanos);
+            final boolean earlierRerouteNeeded;
+            DelayedRerouteTask existingTask = delayedRerouteTask.get();
+            DelayedRerouteTask newTask = new DelayedRerouteTask(nextDelay, currentNanoTime);
+            if (existingTask == null) {
+                earlierRerouteNeeded = true;
+            } else if (newTask.scheduledTimeToRunInNanos() < existingTask.scheduledTimeToRunInNanos()) {
+                // we need an earlier delayed reroute
+                logger.trace("cancelling existing delayed reroute task as delayed reroute has to happen [{}] earlier",
+                    TimeValue.timeValueNanos(existingTask.scheduledTimeToRunInNanos() - newTask.scheduledTimeToRunInNanos()));
+                removeTaskAndCancel();
+                earlierRerouteNeeded = true;
+            } else {
+                earlierRerouteNeeded = false;
+            }
+
+            if (earlierRerouteNeeded) {
+                logger.info("scheduling reroute for delayed shards in [{}] ({} delayed shards)", nextDelay,
+                    UnassignedInfo.getNumberOfDelayedUnassigned(state));
+                newTask.schedule();
+                DelayedRerouteTask currentTask = delayedRerouteTask.getAndSet(newTask);
+                assert existingTask == currentTask || currentTask == null;
```

Comment:
that's OK because of the fact that this run by a single thread, but it will be easier on the eye to use:

```
existingTask.cancel()
```

instead of removeTaskAndCancel()